src/sympc/tensor/share_tensor.py

```diff
@@ -288,18 +320,53 @@ def div(self, y: Union[int, float, torch.Tensor, "ShareTensor"]) -> "ShareTensor
 
         return res
 
-    def __getattr__(self, attr_name: str) -> Any:
-        """Access to tensor attributes.
+    def __getattribute__(self, attr_name: str) -> Any:
+        # Do the forward pass
+        # Implementation similar to CrypTen
+        grad_fn = GRAD_FUNCS.get(attr_name, None)
+        if grad_fn and ShareTensor.AUTOGRAD_IS_ON:
+            return functools.partial(forward, self, grad_fn)
 
-        Args:
-            attr_name (str): Name of the attribute.
+        return object.__getattribute__(self, attr_name)
 
-        Returns:
-            Any: Attribute.
-        """
-        tensor = self.tensor
-        res = getattr(tensor, attr_name)
-        return res
+    def backward(self, grad_input=None) -> Any:
+        if not self.requires_grad:
+            return
+
+        ShareTensor.AUTOGRAD_IS_ON = False
+
+        if grad_input is None:
+            grad_input = ShareTensor(torch.ones_like(self.tensor), session=self.session)
+
+        if self.grad is None:
+            self.grad = grad_input
+        else:
+            self.grad = self.grad + grad_input
+
+        if len(self.parents) == 0:
+            print(
+                f"We can not propagate from this node {self} because it does not have parents"
+            )
+            return
+
+        self.nr_out_edges -= 1
+        if self.nr_out_edges > 0:
+            # For the moment we presume all parents are differentiable
+            print("We will visit this node when all the parents returned the gradients")
+            return
+
+        if self.grad_fn is None:
+            raise ValueError(f"Do not know how to propagate {self}")
+
+        grad = self.grad_fn.backward(self.ctx, self.grad)
+        if not isinstance(grad, (list, tuple)):
+            grad = (grad,)
+
+        for idx, parent in enumerate(self.parents):
+            print(parent)
+            parent.backward(grad_input=grad[idx])
+
```

Comment:
TODO: Clear the context and gradients at the end.
If leaf_node we should leave the gradient